File: local_examples/cmds_hash/redis-py/cmds_hash.py
# EXAMPLE: cmds_hash
# HIDE_START
import redis

r = redis.Redis(host="localhost", port=6379, db=0, decode_responses=True)
# HIDE_END

# STEP_START hdel
hdel1 = r.hset("myhash", "field1", "foo")
print(hdel1)
# >>> 1

hdel2 = r.hget("myhash", "field1")
print(hdel2)
# >>> 1

hdel3 = r.hget("myhash", "field2")
print(hdel3)
# >>> 0

# REMOVE_START
assert hdel1 == 1
assert hdel2 == "foo"
assert hdel3 == None
r.delete("myhash")
# REMOVE_END
# STEP_END

# STEP_START hset
res1 = r.hset("myhash", "field1", "Hello")
print(res1)
# >>> 1

res2 = r.hget("myhash", "field1")
print(res2)
# >>> Hello

res3 = r.hset("myhash", mapping={"field2": "Hi", "field3": "World"})
print(res3)
# >>> 2

res4 = r.hget("myhash", "field2")
print(res4)
# >>> Hi

res5 = r.hget("myhash", "field3")
print(res5)
# >>> World

res6 = r.hgetall("myhash")
print(res6)
# >>> { "field1": "Hello", "field2": "Hi", "field3": "World" }

# REMOVE_START
assert res1 == 1
assert res2 == "Hello"
assert res3 == 2
assert res4 == "Hi"
assert res5 == "World"
assert res6 == { "field1": "Hello", "field2": "Hi", "field3": "World" }
r.delete("myhash")
# REMOVE_END
# STEP_END

# STEP_START hget
res7 = r.hset("myhash", "field1", "foo")
print(res7)
# >>> 1

res8 = r.hget("myhash", "field1")
print(res8)
# >>> foo

res9 = r.hget("myhash", "field2")
print(res9)
# >>> None

# REMOVE_START
assert res7 == 1
assert res8 == "foo"
assert res9 == None
r.delete("myhash")
# REMOVE_END
# STEP_END

# STEP_START hgetall
res10 = r.hset("myhash", mapping={"field1": "Hello", "field2": "World"})

res11 = r.hgetall("myhash")
print(res11) # >>> { "field1": "Hello", "field2": "World" }

# REMOVE_START
assert res11 == { "field1": "Hello", "field2": "World" }
r.delete("myhash")
# REMOVE_END
# STEP_END

# STEP_START hvals
res10 = r.hset("myhash", mapping={"field1": "Hello", "field2": "World"})

res11 = r.hvals("myhash")
print(res11) # >>> [ "Hello", "World" ]

# REMOVE_START
assert res11 == [ "Hello", "World" ]
r.delete("myhash")
# REMOVE_END
# STEP_END

# STEP_START hexpire
# Set up hash with fields
r.hset("myhash", mapping={"field1": "Hello", "field2": "World"})

# Set expiration on hash fields
res12 = r.hexpire("myhash", 10, "field1", "field2")
print(res12)  # >>> [1, 1]

# Check TTL of the fields
res13 = r.httl("myhash", "field1", "field2")
print(res13)  # >>> [10, 10] (or close to 10)

# Try to set expiration on non-existent field
res14 = r.hexpire("myhash", 10, "nonexistent")
print(res14)  # >>> [-2]

# REMOVE_START
assert res12 == [1, 1]
assert all(ttl > 0 for ttl in res13)  # TTL should be positive
assert res14 == [-2]
r.delete("myhash")
# REMOVE_END
# STEP_END

File: local_examples/cmds_hash/ruby/cmds_hash.rb
# EXAMPLE: cmds_hash
# HIDE_START
require 'redis'

r = Redis.new
# HIDE_END

# REMOVE_START
def assert_equal(expected, actual)
  raise "Expected #{expected.inspect}, got #{actual.inspect}" unless actual == expected
end

r.del('myhash')
# REMOVE_END

# STEP_START hdel
res1 = r.hset('myhash', 'field1', 'foo')
puts res1 # >>> 1

res2 = r.hdel('myhash', 'field1')
puts res2 # >>> 1

res3 = r.hdel('myhash', 'field2')
puts res3 # >>> 0
# STEP_END

# REMOVE_START
assert_equal(1, res1)
assert_equal(1, res2)
assert_equal(0, res3)
r.del('myhash')
# REMOVE_END

# STEP_START hset
res4 = r.hset('myhash', 'field1', 'Hello')
puts res4 # >>> 1

res5 = r.hget('myhash', 'field1')
puts res5 # >>> Hello

res6 = r.hset('myhash', { 'field2' => 'Hi', 'field3' => 'World' })
puts res6 # >>> 2

res7 = r.hget('myhash', 'field2')
puts res7 # >>> Hi

res8 = r.hget('myhash', 'field3')
puts res8 # >>> World

res9 = r.hgetall('myhash')
puts res9.inspect
# >>> {"field1"=>"Hello", "field2"=>"Hi", "field3"=>"World"}
# STEP_END

# REMOVE_START
assert_equal(1, res4)
assert_equal('Hello', res5)
assert_equal(2, res6)
assert_equal('Hi', res7)
assert_equal('World', res8)
assert_equal({ 'field1' => 'Hello', 'field2' => 'Hi', 'field3' => 'World' }, res9)
r.del('myhash')
# REMOVE_END

# STEP_START hget
res10 = r.hset('myhash', 'field1', 'foo')
puts res10 # >>> 1

res11 = r.hget('myhash', 'field1')
puts res11 # >>> foo

res12 = r.hget('myhash', 'field2')
puts res12.inspect # >>> nil
# STEP_END

# REMOVE_START
assert_equal(1, res10)
assert_equal('foo', res11)
assert_equal(nil, res12)
r.del('myhash')
# REMOVE_END

# STEP_START hmget
r.hset('myhash', { 'field1' => 'Hello', 'field2' => 'World' })

res13 = r.hmget('myhash', 'field1', 'field2', 'nofield')
puts res13.inspect # >>> ["Hello", "World", nil]
# STEP_END

# REMOVE_START
assert_equal(['Hello', 'World', nil], res13)
r.del('myhash')
# REMOVE_END

# STEP_START hgetall
r.hset('myhash', { 'field1' => 'Hello', 'field2' => 'World' })

res14 = r.hgetall('myhash')
puts res14.inspect # >>> {"field1"=>"Hello", "field2"=>"World"}
# STEP_END

# REMOVE_START
assert_equal({ 'field1' => 'Hello', 'field2' => 'World' }, res14)
r.del('myhash')
# REMOVE_END

# STEP_START hvals
r.hset('myhash', { 'field1' => 'Hello', 'field2' => 'World' })

# HVALS follows the hash's field order, which Redis does not promise, so sort.
res15 = r.hvals('myhash').sort
puts res15.inspect # >>> ["Hello", "World"]
# STEP_END

# REMOVE_START
assert_equal(['Hello', 'World'], res15)
r.del('myhash')
# REMOVE_END

# STEP_START hexpire
# Set up a hash with two fields.
r.hset('myhash', { 'field1' => 'Hello', 'field2' => 'World' })

# Set an expiration on both fields.
res16 = r.hexpire('myhash', 10, 'field1', 'field2')
puts res16.inspect # >>> [1, 1]

# Check the TTL of the fields.
res17 = r.httl('myhash', 'field1', 'field2')
puts res17.inspect # >>> [10, 10]

# Try to set an expiration on a field that does not exist.
res18 = r.hexpire('myhash', 10, 'nonexistent')
puts res18.inspect # >>> [-2]
# STEP_END

# REMOVE_START
assert_equal([1, 1], res16)
assert_equal(true, res17.all? { |ttl| ttl > 0 })
assert_equal([-2], res18)
r.del('myhash')
# REMOVE_END

# STEP_START hlen
res19 = r.hset('myhash', 'field1', 'Hello')
puts res19 # >>> 1

res20 = r.hset('myhash', 'field2', 'World')
puts res20 # >>> 1

res21 = r.hlen('myhash')
puts res21 # >>> 2
# STEP_END

# REMOVE_START
assert_equal(1, res19)
assert_equal(1, res20)
assert_equal(2, res21)
r.del('myhash')
r.close
# REMOVE_END
